lib/grizzly/trace.ex

```diff
@@ -0,0 +1,109 @@
+defmodule Grizzly.Trace do
+  @moduledoc """
+  Module that tracks the commands that are sent and received by Grizzly
+
+  The trace will hold in memory the last 300 messages. If you want to generate
+  a log file of the trace records you use `Grizzly.Trace.dump/1`.
+
+  The log format is:
+
+  ```
+  timestamp source destination sequence_number command_name command_parameters
+  ```
+
+  If you want to list the records that are currently being held in memory you
+  can use `Grizzly.Trace.list/0`.
+
+  If you want to start traces from a fresh start you can call
+  `Grizzly.Trace.clear/0`.
+  """
+
+  use GenServer
+
+  alias Grizzly.Trace.{Record, RecordQueue}
+
+  @type src() :: String.t()
+  @type dest() :: String.t()
+
+  @type log_opt() :: {:src, src()} | {:dest, dest()}
+
+  @doc """
+  Start the trace server
+  """
+  @spec start_link(keyword()) :: GenServer.on_start()
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  @doc """
+  Log the trace information
+  """
+  @spec log(binary(), [log_opt()]) :: :ok
+  def log(binary, opts \\ []) do
+    GenServer.cast(__MODULE__, {:log, binary, opts})
+  end
+
+  @doc """
+  Dump the trace records into a file
+  """
+  @spec dump(Path.t()) :: :ok
+  def dump(file) do
+    GenServer.call(__MODULE__, {:dump, file})
+  end
+
+  @doc """
+  Force clear the records from the trace
+  """
+  @spec clear() :: :ok
+  def clear() do
+    GenServer.call(__MODULE__, :clear)
+  end
+
+  @doc """
+  List all the records currently being traced
+  """
+  @spec list() :: [Record.t()]
+  def list() do
+    GenServer.call(__MODULE__, :list)
+  end
+
+  @impl GenServer
+  def init(_args) do
+    {:ok, RecordQueue.new()}
+  end
+
+  @impl GenServer
+  def handle_cast({:log, binary, opts}, records) do
+    record = Record.new(binary, opts)
+
+    {:noreply, RecordQueue.add_record(records, record)}
```

Comment:
Is there an issue writing to the log on each request? I think that performance will be fine and then memory won't be an issue for a long running trace.

Reply:
We were worried about the file being written indefinitely, so the current code limits the memory usage to 300 items.

I can go back to writing to a file, but do you think there should be some checks to ensure we just don't write too much data to disk? I am thinking bout when a hub runs for weeks at a time. It might not be enough logging to really matter.